taichi/program/program.h

```diff
@@ -24,6 +25,47 @@
 
 TLANG_NAMESPACE_BEGIN
 
+struct JITEvaluatorId {
+  std::thread::id thread_id;
+  // Note that on certain backends (e.g. CUDA), functions created in one
+  // thread cannot be used in another. Hence the thread_id member.
+  int op;
+  DataType ret, lhs, rhs;
+  bool is_binary;
+
+  UnaryOpType unary_op() const {
+    TI_ASSERT(!is_binary);
+    return (UnaryOpType)op;
+  }
+
+  BinaryOpType binary_op() const {
+    TI_ASSERT(is_binary);
+    return (BinaryOpType)op;
+  }
+
+  bool operator==(const JITEvaluatorId &o) const {
+    return thread_id == o.thread_id && op == o.op && ret == o.ret &&
+           lhs == o.lhs && rhs == o.rhs && is_binary == o.is_binary;
+  }
+};
+
+TLANG_NAMESPACE_END
+
+namespace std {
+template <>
+struct hash<taichi::lang::JITEvaluatorId> {
+  std::size_t operator()(taichi::lang::JITEvaluatorId const &id) const
+      noexcept {
+    return (std::size_t)id.op | ((std::size_t)id.ret << 8) |
+           ((std::size_t)id.lhs << 16) | ((std::size_t)id.rhs << 24) |
+           ((std::size_t)id.is_binary << 31) |
+           (std::hash<std::thread::id>{}(id.thread_id) << 32);
+  }
+};
+}  // namespace std
```

Comment:
Cool! But I suggest to use `hash{}(id.op) ^ hash{}(id.ret) ^ ...` for better hashability.

Reply:
Good idea!

Reply:
I would not suggest this because `hash{}(1) ^ hash{}(2) == hash{}(2) ^ hash{}(1)`...

Reply:
https://stackoverflow.com/questions/38304877/why-stdhashint-seems-to-be-identity-function

I think it's better to hand-write hash functions for the variables we know to be very small (but `std::hash` doesn't know).

Reply:
I guess when probably combined with `<<` (our case), bit xors make more sense than bit ors?

Reply:
Or maybe we can use `std::hash(bit concatenation of id.op, id.ret, ...)`.

Reply:
> I guess when properly combined with `<<` (our case), bit xors make more sense than bit ors?

Oh sure. `hash{}(id.op) ^ hash{}(id.ret << 8)` SGTM.

Reply:
Now I actually think
```C++
return ((std::size_t)id.op | ((std::size_t)id.ret << 8) |
            ((std::size_t)id.lhs << 16) | ((std::size_t)id.rhs << 24) |
            ((std::size_t)id.is_binary << 31)) ^ // Note: xor here
           (std::hash<std::thread::id>{}(id.thread_id) << 32);
```
is good enough, given now we know the number of bits of these fields. WDYT?

Reply:
> Now I actually think
> 
> ```c++
> return ((std::size_t)id.op | ((std::size_t)id.ret << 8) |
>             ((std::size_t)id.lhs << 16) | ((std::size_t)id.rhs << 24) |
>             ((std::size_t)id.is_binary << 31)) ^ // Note: xor here
>            (std::hash<std::thread::id>{}(id.thread_id) << 32);
> ```
> 
> is good enough, given now we know the number of bits of these fields. WDYT?

I agree. The expression before `^` here is actually a perfect hash function.